src/communication/channel.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "channel.hpp"
+#include "log.hpp"
+#include "utils/aosheader.hpp"
+#include "utils/checksum.hpp"
+
+namespace aos::zephyr::communication {
+
+Mutex Channel::sWriteMutex;
+
+Channel::Channel(TransportItf* transport, int port)
+    : mTransport(transport)
+    , mPort(port)
+{
+}
+
+Error Channel::Connect()
+{
+    UniqueLock lock(mMutex);
+
+    LOG_DBG() << "Connect channel port=" << mPort;
+
+    if (mTransport->IsOpened()) {
+        return ErrorEnum::eNone;
+    }
+
+    return ErrorEnum::eRuntime;
+}
+
+Error Channel::Close()
+{
+    UniqueLock lock(mMutex);
+
+    LOG_DBG() << "Close channel port=" << mPort;
+
+    mClose = true;
+    mCondVar.NotifyAll();
+
+    return ErrorEnum::eNone;
+}
+
+int Channel::Read(void* buffer, size_t size)
+{
+    UniqueLock lock(mMutex);
+
+    LOG_DBG() << "Read channel port=" << mPort << " size=" << size;
+
+    mBufferLen = 0;
+
+    while (mBufferLen < size) {
+        mReadReady = true;
+        mBuffer    = reinterpret_cast<uint8_t*>(buffer) + mBufferLen;
+        mBufferLen = size - mBufferLen;
+
+        mCondVar.NotifyAll();
+
+        mCondVar.Wait(lock, [this] { return !mReadReady || mClose; });
+
+        if (mClose) {
+            return -1;
+        }
+    }
+
+    return static_cast<int>(size);
+}
+
+int Channel::Write(const void* buffer, size_t size)
+{
+    auto header = PrepareHeader(mPort, Array<uint8_t>(reinterpret_cast<const uint8_t*>(buffer), size));
```

Comment:
Better to send header in channel manager. Also, static mutex should belong to channel manager in this case.

Reply:
Done